icu4c/source/common/mlbe.cpp

```diff
@@ -128,135 +122,169 @@ int32_t MlBreakEngine::divideUpRange(UText *inText, int32_t rangeStart, int32_t
-        score += (2 * fModel.geti(feature.setTo(u"TW3:", 4)
-                                      .append(elementList[2])
-                                      .append(elementList[3])
-                                      .append(elementList[4])));
+    if (indexList[2] != -1 && indexList[3] != -1 && indexList[4] != -1) {
+        // TW3
+        end = (indexList[5] != -1) ? indexList[5] : codeUts;
+        score += fModel[11].geti(inString.tempSubString(indexList[2], end - indexList[2]));
     }
-    if (elementList[3] != INVALID && elementList[4] != INVALID && elementList[5] != INVALID) {
-        score += (2 * fModel.geti(feature.setTo(u"TW4:", 4)
-                                      .append(elementList[3])
-                                      .append(elementList[4])
-                                      .append(elementList[5])));
+    if (indexList[3] != -1 && indexList[4] != -1 && indexList[5] != -1) {
+        // TW4
+        end = codeUts;
+        score += fModel[12].geti(inString.tempSubString(indexList[3], end - indexList[3]));
     }
     if (score > 0) {
         boundary.addElement(index, status);
         numBreaks++;
     }
 }
 
-int32_t MlBreakEngine::initElementList(const UnicodeString &inString, UChar32* elementList,
-                                         UErrorCode &status) const {
+int32_t MlBreakEngine::initIndexList(const UnicodeString &inString, int32_t* indexList,
+                                     UErrorCode &status) const {
     if (U_FAILURE(status)) {
         return 0;
     }
     int32_t index = 0;
     int32_t length = inString.countChar32();
-    UChar32 w1, w2, w3, w4, w5, w6;
-    w1 = w2 = w3 = w4 = w5 = w6 = INVALID;
+    int32_t i3, i4, i5, i6;
+    i3 = i4 = i5 = i6 = -1;
     if (length > 0) {
-        w3 = inString.char32At(0);
-        index += U16_LENGTH(w3);
+        i3 = 0;
+        index += U16_LENGTH(inString.char32At(0));
         if (length > 1) {
-            w4 = inString.char32At(index);
-            index += U16_LENGTH(w4);
+            i4 = index;
+            index += U16_LENGTH(inString.char32At(index));
             if (length > 2) {
-                w5 = inString.char32At(index);
-                index += U16_LENGTH(w5);
+                i5 = index;
+                index += U16_LENGTH(inString.char32At(index));
                 if (length > 3) {
-                    w6 = inString.char32At(index);
-                    index += U16_LENGTH(w6);
+                    i6 = index;
+                    index += U16_LENGTH(inString.char32At(index));
                 }
             }
         }
     }
-    elementList[0] = w1;
-    elementList[1] = w2;
-    elementList[2] = w3;
-    elementList[3] = w4;
-    elementList[4] = w5;
-    elementList[5] = w6;
-
+    indexList[0] = -1;
+    indexList[1] = -1;
+    indexList[2] = i3;
+    indexList[3] = i4;
+    indexList[4] = i5;
+    indexList[5] = i6;
     return index;
 }
 
 void MlBreakEngine::loadMLModel(UErrorCode &error) {
```

Comment:
Could we add a note about where are the code in BudouX to generate the model stored in data/brkitr/adaboost/jaml.txt here A URL pointing to the exact tool that anyone in addition to the PR author or BudouX (such as me) can run to recreate that file. I am concern about the maintainability of that file.
It is now looks like a voodoo magic to me that we have no idea how to modify it.

Reply:
OK, I'll create another PR to upload the tool.